### PR TITLE
LUCENE-10608: Implement Weight#count on pure conjunctions.

### PR DESCRIPTION
**Pull request by jpountz**
https://issues.apache.org/jira/browse/LUCENE-10608

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java [hunk 1]
```diff
@@ -344,6 +344,45 @@ public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
...
+    // or all clauses but one match all docs.
+    if (weightedClauses.isEmpty()) {
+      return 0;
+    }
+    for (WeightedBooleanClause weightedClause : weightedClauses) {
+      switch (weightedClause.clause.getOccur()) {
+        case FILTER:
+        case MUST:
+          break;
+        case MUST_NOT:
```
Comment (zhaih) [inline lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java:359]: Why can't we include the MUST_NOT clause here? Isn't the count effectively `maxDoc - clause.count()`?

File: lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java [hunk 2]
```diff
@@ -344,6 +344,45 @@ public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
...
+        case SHOULD:
```
Comment (LuXugang) [inline lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java:360]: If `weightedClauses.size() == 1`, and this clause's occur is `SHOULD`, we could also leverage `Weight#count`, should we need to specialize this case?

File: lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java (hunk 1 above)
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java:359]: You are right. I think there are a few more cases we could optimize, like pure disjunctions of term queries on a single-valued field where we could just sum up counts. I was thinking of introducing these optimizations one at a time to keep changes easy to review and test.

File: lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java (hunk 2 above)
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java:360]: It should already be handled by the fact that `BooleanQuery#rewrite` rewrites single-clause queries.

**Review by zhaih — APPROVED** [review]

**Comment by jpountz** [comment]
Thanks @zhaih !

**Pull request merged**